ddl/partition.go

```diff
@@ -33,16 +36,239 @@ import (
 	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/meta"
 	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
 	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
 )
 
 const (
 	partitionMaxValue = "MAXVALUE"
 )
 
+func checkAddPartition(t *meta.Meta, job *model.Job) (*model.TableInfo, *model.PartitionInfo, []model.PartitionDefinition, error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, errors.Trace(err)
+	}
+	partInfo := &model.PartitionInfo{}
+	err = job.DecodeArgs(&partInfo)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, nil, errors.Trace(err)
+	}
+	if len(tblInfo.Partition.AddingDefinitions) > 0 {
+		return tblInfo, partInfo, tblInfo.Partition.AddingDefinitions, nil
+	}
+	return tblInfo, partInfo, []model.PartitionDefinition{}, nil
+}
+
+func onAddTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	// Handle the rolling back job
+	if job.IsRollingback() {
+		ver, err := onDropTablePartition(t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	tblInfo, partInfo, addingDefinitions, err := checkAddPartition(t, job)
+	if err != nil {
+		return ver, err
+	}
+
+	// addingDefinition's len is 0 meaning the job is in the initial state of add partition.
+	// Here should use partInfo from job directly and do some check action.
+	if len(addingDefinitions) == 0 {
+		err = checkAddPartitionTooManyPartitions(uint64(len(tblInfo.Partition.Definitions) + len(partInfo.Definitions)))
```

Comment:
move line 89 ~ 105 after line 111, simply use `job.SchemaState == model.StateNone` to instead of the initial state of add partition?

Reply:
good